core/sessions/ldapauth/reaper.go

```diff
@@ -90,43 +94,129 @@ func (ldSync *LDAPServerStateSyncer) Work() {
+	// Dedupe preserving order of highest role (sorted)
+	// Preserve members as a map for future lookup
+	upstreamUserStateMap := make(map[string]sessions.User)
 	for _, user := range users {
-		if _, ok := uniqueRef[user.Email]; !ok {
-			uniqueRef[user.Email] = struct{}{}
-			upstreamUserState = append(upstreamUserState, user)
+		if _, ok := upstreamUserStateMap[user.Email]; !ok {
+			upstreamUserStateMap[user.Email] = user
 		}
 	}
 
-	// upstreamUserState is now the most up to date source of truth
-	// Update state of local ldap tables, purging users not present in
-	// up to date list, and downgrading roles where applicable
+	// upstreamUserStateMap is now the most up to date source of truth
+	// Now sync database sessions and roles with new data
 	err = ldSync.q.Transaction(func(tx pg.Queryer) error {
-		emailList := []interface{}{}
-		for _, user := range upstreamUserState {
-			emailList = append(emailList, user.Email)
+		// First, purge users present in the local ldap_sessions table but not in the upstream server
+		type LDAPSession struct {
+			UserEmail string
+			UserRole  sessions.UserRole
+		}
+		var existingSessions []LDAPSession
+		if err := tx.Select(&existingSessions, "SELECT user_email, user_role FROM ldap_sessions"); err != nil {
+			return errors.Wrap(err, "Unable to query ldap_sessions table")
+		}
+		var existingAPITokens []LDAPSession
+		if err := tx.Select(&existingAPITokens, "SELECT user_email, user_role FROM ldap_user_api_tokens"); err != nil {
+			return errors.Wrap(err, "Unable to query ldap_user_api_tokens table")
+		}
+
+		// Create existing sessions and API tokens lookup map for later
+		existingSessionsMap := make(map[string]LDAPSession)
+		for _, sess := range existingSessions {
+			existingSessionsMap[sess.UserEmail] = sess
+		}
+		existingAPITokensMap := make(map[string]LDAPSession)
+		for _, sess := range existingAPITokens {
+			existingAPITokensMap[sess.UserEmail] = sess
+		}
+
+		// Populate list of session emails present in the local session table but not in the upstream state
+		emailsToPurge := []interface{}{}
+		for _, ldapSession := range existingSessions {
+			if _, ok := upstreamUserStateMap[ldapSession.UserEmail]; !ok {
+				emailsToPurge = append(emailsToPurge, ldapSession.UserEmail)
+			}
+		}
+		// Likewise for API Tokens table
+		apiTokenEmailsToPurge := []interface{}{}
+		for _, ldapSession := range existingAPITokens {
+			if _, ok := upstreamUserStateMap[ldapSession.UserEmail]; !ok {
+				apiTokenEmailsToPurge = append(apiTokenEmailsToPurge, ldapSession.UserEmail)
+			}
+		}
+
+		// Remove any active sessions this user may have
+		if len(emailsToPurge) > 0 {
+			placeholders := make([]string, len(emailsToPurge))
+			for i := range emailsToPurge {
+				placeholders[i] = fmt.Sprintf("$%d", i+1)
+			}
+			query := fmt.Sprintf("DELETE FROM ldap_sessions WHERE user_email IN (%s)", strings.Join(placeholders, ", "))
+			_, err := ldSync.q.Exec(query, emailsToPurge...)
```

Comment:
You don't need to construct this whole `placeholders` list. Try this instead:
```
import "github.com/lib/pq"
...
	_, err := ldSync.q.Exec("DELETE FROM ldap_sessions WHERE user_email = ANY($1)", pq.Array(emailsToPurge))
```

Reply:
Oh that's really nice, thanks! Updated now